core/cachemap.go

```diff
@@ -0,0 +1,89 @@
+package core
+
+import (
+	"sync"
+)
+
+type cacheMap[K comparable, T any] struct {
+	cache map[K]*cacheInitializer[K, T]
+	l     sync.Mutex
+}
+
+func newCacheMap[K comparable, T any]() *cacheMap[K, T] {
+	return &cacheMap[K, T]{
+		cache: make(map[K]*cacheInitializer[K, T]),
+	}
+}
+
+func (cache *cacheMap[K, T]) Get(key K) (T, bool) {
+	cache.l.Lock()
+	defer cache.l.Unlock()
+
+	init, found := cache.cache[key]
+	if found {
+		init.Wait()
+		return init.val, true
+	}
+
+	var zero T
+	return zero, false
+}
+
+type Initializer[T any] interface {
+	// Put sets the value and wakes up any callers waiting for it.
+	Put(val T)
+
+	// Release removes the initializer from the cache if it has not been
+	// initialized.
+	Release()
+}
+
+func (cache *cacheMap[K, T]) GetOrInitialize(key K) (T, Initializer[T], bool) {
+	cache.l.Lock()
+	defer cache.l.Unlock()
+
+	init, found := cache.cache[key]
+	if found {
+		init.Wait()
+		return init.val, nil, true
```

Comment:
@sipsma Ready for another look. :pray: 

I switched to something closer to singleflight's callback flow, the difference being that errors won't be cached forever, but they will still be returned to all concurrent runs.

To be honest this is all theoretical. It felt a little aggressive to cache errors forever since they could be from network flakes, but realistically any retries might be done with an entirely different session anyway. :man_shrugging:

Reply:
Yeah that makes perfect sense I think, it should be the same as what the caching behavior of buildkit is essentially